octobot_trading/orders/order_state.py

```diff
@@ -0,0 +1,184 @@
+                               f" on {self.order.exchange_manager.exchange_name}")
+
+    async def initialize_impl(self) -> None:
+        """
+        Default OrderState initialization process
+        """
+        await self.update()
+
+    async def terminate(self) -> None:
+        """
+        Implement the state ending process
+        Can be portfolio updates, fees request, linked order cancellation, Trade creation etc...
+        """
+        raise NotImplementedError("terminate not implemented")
+
+    async def update(self) -> None:
+        """
+        Update the order state
+        Try to fix the pending state or terminate
+        """
+        if self.is_pending() and self.state is not OrderStates.REFRESHING and self.updating_task is None:
+            await self.synchronize()
+        else:
+            await self.terminate()
+
+    async def synchronize(self) -> None:
+        """
+        Implement the exchange synchronization process
+        Should begin by setting the state to REFRESHING
+        Should end by :
+        - calling terminate if the state is terminated
+        - restoring the initial state if nothing has been changed with synchronization or if sync failed
+        """
+        await self._synchronize_order_with_exchange()
+
+    async def _refresh_order_from_exchange(self) -> bool:
+        """
+        Ask OrdersChannel Internal producer to refresh the order from the exchange
+        :return: the result of OrdersProducer.update_order_from_exchange()
+        """
+        return (await get_chan(ORDERS_CHANNEL, self.order.exchange_manager.id).get_internal_producer().
+                update_order_from_exchange(self.order))
+
+    async def on_order_refresh_successful(self):
+        """
+        Called when _synchronize_order_with_exchange succeed to update the order
+        """
+        raise NotImplementedError("_on_order_refresh_successful not implemented")
+
+    async def _synchronize_order_with_exchange(self, retry_on_fail=True,
+                                               retry_timeout=DEFAULT_SYNC_RETRY_TIMEOUT):
+        """
+        Ask the exchange to update the order
+        Also manage the order state during the refreshing process
+        :param retry_on_fail: if the synchronization process should be retried after failure
+        :param retry_timeout: the retry timeout
+        """
+        previous_state = self.state
+        self.state = OrderStates.REFRESHING
+        if await self._refresh_order_from_exchange():
+            try:
+                return await self.on_order_refresh_successful()
+            except Exception as e:
+                self.get_logger().warning(f"Error during order synchronization process : {e}, restoring previous...")
+                self.state = previous_state
+        else:
+            self.state = previous_state
+        if retry_on_fail:
+            self.updating_task = asyncio.create_task(self.postpone_synchronization(timeout=retry_timeout))
```

Comment:
do we really want to lose the current context here and start `postpone_synchronization` in a new task ?

Reply:
Yes I forgot await asyncio.sleep() wasn't a thread block call. :+1: